src/main/java/Tree/RedBlackTree.java

```diff
@@ -0,0 +1,372 @@
+                }
+                parent.setColour(true);
+                gp.setColour(false);
+                rightRotation(gp);
+            }
+            else if (parent.getKey() >  gp.getKey()) {
+                if (node.getKey() < parent.getKey()) {
+                    rightRotation(parent);
+                }
+                parent.setColour(true);
+                gp.setColour(false);
+                leftRotation(gp);
+            }
+        }
+        RedBlackNode<T> rt = (RedBlackNode<T>) root;
+        rt.setColour(true);
+        root = rt;
+    }
+
+    public void delete(int delete_key) {
+        RedBlackNode<T> deleteNode = (RedBlackNode<T>) getNode(delete_key);
+        RedBlackNode<T> nodeToUpdate = deleteNode;
+
+        if (deleteNode == null) return;
+
+        RedBlackNode<T> successor = (RedBlackNode<T>) getSuccessor(deleteNode);
+        if (successor == null) {
+            RedBlackNode<T> predecessor = (RedBlackNode<T>) getPredecessor(deleteNode);
+            if (predecessor == null) {
+                // success and predecessor do not exist
+
+                RedBlackNode<T> parent_node = (RedBlackNode<T>) deleteNode.getParentNode();
+                if (parent_node == null) {
+                    root = null;
+                    current = null;
+                } else {
+                    int parent_key = parent_node.getKey();
+                    if (delete_key < parent_key) {
+                        parent_node.setLeftNode(null);
+                    } else {
+                        parent_node.setRightNode(null);
+                    }
+
+                    nodeToUpdate = deleteNode;
+                    // If deleting node is a leaf and black then mark it as double black
+                    if (nodeToUpdate.getColour()) {
+                        nodeToUpdate.setDoubleBlack(true);
+                    }
+                }
+            } else {
+                swapWithPredecessor(deleteNode, predecessor);
+                nodeToUpdate = (RedBlackNode<T>) predecessor.getParentNode();
+                if (predecessor.getColour() && predecessor.getRightNode() == null && predecessor.getLeftNode() == null) {
+                    handleDoubleBlack(predecessor);
+                }
+            }
+        } else {
+            swapWithSuccessor(deleteNode, successor);
+            nodeToUpdate = (RedBlackNode<T>) successor.getParentNode();
+            if (successor.getColour() && successor.getLeftNode() == null && successor.getRightNode() == null) {
+                handleDoubleBlack(successor);
+            }
+        }
+
+        if (nodeToUpdate != null) {
+            adjustAfterDelete(nodeToUpdate);
+        }
+    }
+
+    public void adjustAfterDelete(RedBlackNode<T> node) {
+        if (node.isDoubleBlack()) {
+            handleDoubleBlack(node);
+        } else if (node.getColour()) {
+            deleteCase1(node);
+        }
+    }
+
+    public void handleDoubleBlack(RedBlackNode<T> node) {
```

Comment:
For some of the functions that is hard to understand the context,
please leave comments on what they are supposed to do.